azurerm/resource_arm_virtual_machine_data_disk_attachment.go

```diff
@@ -0,0 +1,287 @@
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+				ConflictsWith:    []string{"vhd_uri"},
+			},
+
+			"managed_disk_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(compute.PremiumLRS),
+					string(compute.StandardLRS),
+				}, true),
+				ConflictsWith: []string{"vhd_uri"},
+			},
+
+			"caching": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"disk_size_gb": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validateDiskSizeGB,
+			},
+		},
+	}
+}
+
+func resourceArmVirtualMachineDataDiskAttachmentCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).vmClient
+	ctx := meta.(*ArmClient).StopContext
+
+	virtualMachineId := d.Get("virtual_machine_id").(string)
+	parsedVirtualMachineId, err := parseAzureResourceID(virtualMachineId)
+	if err != nil {
+		return err
+	}
+	resourceGroup := parsedVirtualMachineId.ResourceGroup
+	virtualMachineName := parsedVirtualMachineId.Path["virtualMachines"]
+
+	azureRMLockByName(virtualMachineName, virtualMachineResourceName)
+	defer azureRMUnlockByName(virtualMachineName, virtualMachineResourceName)
+
+	virtualMachine, err := client.Get(ctx, resourceGroup, virtualMachineName, compute.InstanceView)
+	if err != nil {
+		if utils.ResponseWasNotFound(virtualMachine.Response) {
+			return fmt.Errorf("Virtual Machine %q (Resource Group %q) was not found", virtualMachineName, resourceGroup)
+		}
+
+		return fmt.Errorf("Error loading Virtual Machine %q (Resource Group %q): %+v", virtualMachineName, resourceGroup, err)
+	}
+
+	name := d.Get("name").(string)
+	createOption := d.Get("create_option").(string)
+	lun := int32(d.Get("lun").(int))
+
+	expandedDisk := compute.DataDisk{
+		Name:         utils.String(name),
+		Lun:          utils.Int32(lun),
+		CreateOption: compute.DiskCreateOptionTypes(createOption),
+	}
+	if v, ok := d.GetOk("vhd_uri"); ok {
+		expandedDisk.Vhd = &compute.VirtualHardDisk{
+			URI: utils.String(v.(string)),
+		}
+	} else {
+		storageAccountType := d.Get("managed_disk_type").(string)
+		expandedDisk.ManagedDisk = &compute.ManagedDiskParameters{
+			StorageAccountType: compute.StorageAccountTypes(storageAccountType),
+		}
+
+		if v, ok := d.GetOk("managed_disk_id"); ok {
+			expandedDisk.ManagedDisk.ID = utils.String(v.(string))
+		}
+	}
+
+	if v, ok := d.GetOk("caching"); ok {
+		expandedDisk.Caching = compute.CachingTypes(v.(string))
+	}
+
+	if v, ok := d.GetOk("disk_size_gb"); ok {
+		expandedDisk.DiskSizeGB = utils.Int32(int32(v.(int)))
+	}
+
+	disks := *virtualMachine.StorageProfile.DataDisks
+	if !d.IsNewResource() {
+		// find the existing disk, remove it from the array
+		dataDisks := make([]compute.DataDisk, 0)
+		for _, dataDisk := range disks {
+			if !strings.EqualFold(*dataDisk.Name, *expandedDisk.Name) {
+				disks = append(disks, dataDisk)
```

Comment:
It seems that dataDisks is never used since initialization. I guess you were using it store the left disks. Then I think you should do:

dataDisks = append(dataDisks, dataDisk)